### PR TITLE
bigvm: Replace str-keyed host_fractions with ratio

### PR DESCRIPTION
**Pull request by grandchild**
Before, the filter for placing big VMs supported a fixed configuration
dict of hypervisor cache to big VM memory size ratios, identified by
string keys. This became cumbersome if the list of ratios got longer.

Repurpose the 'host_fraction' extra spec parameter for big VMs to
support a float number (or comma-separated list of numbers) between 0
and 1 expressing one or more possible ratios of VM/hypervisor memory.

### CODE REVIEW

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py [hunk 1]
```diff
@@ -231,141 +232,110 @@ def call(a, b):
...
     def test_extra_specs_half_positive(self):
         """test specified half size"""
-        CONF.set_override('bigvm_host_size_filter_uses_flavor_extra_specs',
-                          True, 'filter_scheduler')
-        CONF.set_override('bigvm_host_size_filter_host_fractions',
-                          {'full': '1', 'half': '0.5'}, 'filter_scheduler')
         spec_obj = objects.RequestSpec(
             flavor=objects.Flavor(memory_mb=CONF.bigvm_mb,
-                                  extra_specs={'host_fraction': 'full,half'},
+                                  extra_specs={'host_fraction': '.5, 1'},
```
Comment (joker-official) [inline nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: Are both space and missing `0` here done on purpose? If yes, I'd rather test them in another test-function, that explicitly states that it's testing different `host_fraction` formats.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py [hunk 2]
```diff
@@ -231,141 +232,110 @@ def call(a, b):
...
+        self.filt_cls._HV_SIZE_CACHE[host.uuid] = CONF.bigvm_mb * 2.5 + 1024
+        self.assertTrue(self.filt_cls.host_passes(host, spec_obj))
+
+        spec_obj.flavor.extra_specs['host_fraction'] = '0.6'
+        self.filt_cls._HV_SIZE_CACHE[host.uuid] = CONF.bigvm_mb * 1.67 + 1024
+        self.assertTrue(self.filt_cls.host_passes(host, spec_obj))
+
+        spec_obj.flavor.extra_specs['host_fraction'] = '0.8'
+        self.filt_cls._HV_SIZE_CACHE[host.uuid] = CONF.bigvm_mb * 1.25 + 1024
+        self.assertTrue(self.filt_cls.host_passes(host, spec_obj))
```
Comment (joker-official) [inline nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: I don't really see a value in those test-cases. They don't seem to test any edge-case really. Can you elaborate on your intention here? I would understand, if you'd test that really all values from the `host_fraction` setting can be used, by adding all those fractions into the flavor at once and then test it against different host sizes, but can't see the point in doing it like this.

File: nova/scheduler/filters/bigvm_filter.py [hunk 3]
```diff
@@ -152,49 +136,36 @@ def _check_flavor_extra_specs(self, host_state, flavor,
...
-                        'supported value (%(specs_values)s). Cannot schedule '
-                        'on %(host_state)s.',
+        host_fractions = set()
+        for x in extra_specs[self._EXTRA_SPECS_KEY].split(','):
+            try:
+                ratio = float(x)
+                if 0.0 <= ratio <= 1.0:
+                    host_fractions.add(ratio * hypervisor_ram_mb)
+            except ValueError:
+                pass
```
Comment (joker-official) [inline nova/scheduler/filters/bigvm_filter.py]: I'd write this as
```python
try:
  ratio = float(x)
except ValueError:
  continue
if 0 <= ratio <= 1:
  host_fractions.add(ratio * hypervisor_ram_mb)
```
to keep the section between try-except as short as necessary.

File: nova/scheduler/filters/bigvm_filter.py [hunk 4]
```diff
@@ -210,16 +181,15 @@ def host_passes(self, host_state, spec_obj):
...
-                                                  spec_obj.flavor,
-                                                  supported_fractions,
-                                                  requested_ram_mb)
-        else:
-            return self._check_no_flavor_extra_specs(host_state,
-                                                     supported_fractions,
-                                                     requested_ram_mb)
+        try:
+            spec_obj.flavor.extra_specs[self._EXTRA_SPECS_KEY]
+        except (KeyError, AttributeError):
```
[the remark this answers is not in the document]
Comment (joker-official) [reply nova/scheduler/filters/bigvm_filter.py]: Given that, I would rather go with `hasattr(spec_obj.flavor, 'extra_specs')` here, if you really think that might happen outside of test-cases.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py [hunk 5]
```diff
@@ -119,7 +120,7 @@ def test_big_vm_without_hv_size(self):
         spec_obj = objects.RequestSpec(
             flavor=objects.Flavor(memory_mb=CONF.bigvm_mb, extra_specs={}))
         host = fakes.FakeHostState('host1', 'compute',
-                {'uuid': uuidsentinel.host1})
+                                   {'uuid': uuidsentinel.host1})
```
[the remark this answers is not in the document]
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: I will change it back here, but in my opinion it is inconsistent with all the other places with multiline parameters (where the second and subsequent lines match the indentation of the opening bracket).

[the remark this answers is not in the document]
Comment (joker-official) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: But I think it was consistent in this file, which is all that matters :D

[the remark this answers is not in the document]
Comment (joker-official) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: Additionally, my point was: you cannot just change on thing and not the other. You introduced inconsistencies in the same file in a functional commit, where the functional changes did not touch this line. If you want it to look different, please create another commit for changing the style.

File: nova/scheduler/filters/bigvm_filter.py (hunk 3 above)
Comment (grandchild) [reply nova/scheduler/filters/bigvm_filter.py]: The working replacement would be:
```python
ratio = None
try:
  ratio = float(x)
except ValueError:
  continue
if 0 <= ratio <= 1:
  host_fractions.add(ratio * hypervisor_ram_mb)
```
Is that also better? I disliked the initial variable declaration, but I don't care either way.

Comment (joker-official) [reply nova/scheduler/filters/bigvm_filter.py]: Why is the initial declaration necessary?

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 1 above)
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: Yes on purpose.
I will consolidate it (`'0.5,1'`), because I don't really expect `float()` to change behaviour anytime soon.

File: nova/scheduler/filters/bigvm_filter.py (hunk 3 above)
Comment (grandchild) [reply nova/scheduler/filters/bigvm_filter.py]: Ah I overlooked the `pass` needs to be replaced by `continue` that's why I got `UnboundLocalError`... My bad, will be fixed. I initially wrote it the way you suggested, but thought that Python didn't expose vars outside of `try`s after all — wrongly so.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 2 above)
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: I considered it valuable to have test-cases with other-than-0.5-or-1 values for ratios, in case a specific mostly-used list emerges through use in production, which might then be reflected in future code by never checking different values than the ones customarily used.

I agree they're not strictly necessary and these values are a bit arbitrary themselves. There's evidently no method for checking _all_ values, but a `range(0, 1, 0.2)` seemed like a compromise. Can remove it... But personally would leave it in. I can change it to add all values at once.

File: nova/scheduler/filters/bigvm_filter.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (grandchild) [reply nova/scheduler/filters/bigvm_filter.py]: You're right, this is superfluous because of the check in `_check_flavor_extra_spec()`. Will remove the whole check.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py [hunk 6]
```diff
@@ -231,141 +232,110 @@ def call(a, b):
...
+        """If the extra_specs flavor attr is missing, we fail."""
         spec_obj = objects.RequestSpec(
-            flavor=objects.Flavor(memory_mb=CONF.bigvm_mb, extra_specs={}))
+            flavor=objects.Flavor(memory_mb=CONF.bigvm_mb))
         host = fakes.FakeHostState('host1', 'compute',
-                {'uuid': uuidsentinel.host1})
-        self.assertTrue(self.filt_cls.host_passes(host, spec_obj))
+                                   {'uuid': uuidsentinel.host1})
+        self.assertRaises(ObjectActionError, self.filt_cls.host_passes,
+                          host, spec_obj)
```
[the remark this answers is not in the document]
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: Honestly, I don't know if this might _reasonably_ happen or not. It seemed like a valid thing to check for. I will see if I can find out whether this seems at all plausible by looking for calling code...

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 2 above)
Comment (joker-official) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: If I understand you correctly, these tests are supposed to guard against us basically hardcoding values in the algorithm. Then imho we should use a random number here and compute the fitting host-fraction in the test.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: But if an exception means the whole scheduler dies, then this needs to be caught (if it's conceivable) or this test removed (if it's not likely to occur).

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 2 above)
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: I thought about random numbers, but thought they were a bit of a smell in tests, because then one test run might fail and the next succeed. We could pre-seed the RNG, though.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: I thought I changed all occurrences.

Nevertheless, you're right that it does warrant a separate commit.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (joker-official) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: This looks helpful:
https://github.com/sapcc/nova/blob/ed2db87278b102a8be34151bccaab81458b96b1c/nova/scheduler/filters/aggregate_instance_extra_specs.py#L45-L49

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 2 above)
Comment (joker-official) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: You can compute the expected host-memory necessary for that random number and set it. What this would check, is that any random number passes the algorithm which means there's an algorithm and no hardcoded map or something. I don't really think that's test-worthy, but who knows.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: Yes doing it this way would consolidate it with that class. I will do that in a separate commit, sometime later... For now I'll just remove the check in `host_passes()` -> smaller diff.

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 2 above)
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: Okay, after some discussion I will remove the test:
1. It tests for programmer maliciousness more than anything. It checks against hardcoded values in the algorithgm, which are unlikely to be introduced by anyone mildly competent.
1. It doesn't need to test the parseability of various numbers. Python-`float()` supposedly works fine.
1. Less code

File: nova/tests/unit/scheduler/filters/test_bigvm_filters.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (grandchild) [reply nova/tests/unit/scheduler/filters/test_bigvm_filters.py]: Removed the whole test.

**Comment by grandchild** [comment]
Sorry for the push-noise. Should be ready now.
